### PR TITLE
OrderBook: market maker network fee

### PR DESCRIPTION
**Pull request by wer1st**
Introduce the custom reduced network fee for maket makers.
`place_limit_order` the following network fees:
- reduced fee if the limit order successfully placed
- regular fee if some error occurs
- regular fee if the limit order was converted into a market order

### CODE REVIEW

File: runtime/src/xor_fee_impls.rs [hunk 1]
```diff
@@ -193,6 +242,9 @@ impl xor_fee::ApplyCustomFees<RuntimeCall, AccountId> for CustomFees {
                     return false;
                 }
             }
+
+            #[cfg(feature = "wip")] // order-book
+            RuntimeCall::OrderBook(order_book::Call::place_limit_order { .. }) => true,
```
[the remark this answers is not in the document]
Comment (wer1st) [reply runtime/src/xor_fee_impls.rs:247]: Yes, just forgot to delete

**Review by bragov4ik — CHANGES_REQUESTED** [review]
Not sure about computation error part
LGTM other than this

File: pallets/order-book/src/fee_calculator.rs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+        let life_ratio = FixedWrapper::from(lifetime) / FixedWrapper::from(max_lifetime);
+
+        let const_part = (FixedWrapper::from(market_maker_max_fee) / FixedWrapper::from(7))
+            * FixedWrapper::from(4);
+        let dynamic_part = (FixedWrapper::from(market_maker_max_fee) / FixedWrapper::from(7))
+            * FixedWrapper::from(3)
+            * life_ratio;
+
+        (const_part + dynamic_part).try_into_balance().ok()
```
[the remark this answers is not in the document]
Comment (wer1st) [reply pallets/order-book/src/fee_calculator.rs:80]: It is better to return `None`

**Review by bragov4ik — APPROVED** [review]

File: pallets/order-book/src/fee_calculator.rs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    /// The idea is to provide the reduced fee for market maker.
+    /// If extrinsic is successfull, user pays maximum half network fee.
+    /// If extrinsic failed, user pays full network fee.
+    ///
+    /// The actual fee contains two parts: constant and dynamic.
+    /// Dynamic part depends on limit order lifetime. The longer the lifetime, the higher the dynamic fee.
+    ///
+    /// const part = (4 / 7) * (base fee / 2)
+    /// dynamic part = (3 / 7) * (base fee / 2) * (lifetime / max_lifetime)
+    pub fn place_limit_order_fee(
```
[the remark this answers is not in the document]
Comment (wer1st) [reply pallets/order-book/src/fee_calculator.rs:53]: It is better to have a type (struct) instead of standalone fn

**Review by bragov4ik — APPROVED** [review]

**Review by vovac12 — APPROVED** [review]

**Pull request merged**